mm2src/coins/eth/v2_activation.rs

```diff
@@ -0,0 +1,276 @@
+                "fallback_swap_contract can't be zero address".to_string(),
+            )
+            .into());
+        }
+    }
+
+    let key_pair: KeyPair =
+        KeyPair::from_secret_slice(priv_key).map_err(|e| EthActivationV2Error::InternalError(e.to_string()))?;
+    let my_address = checksum_address(&format!("{:02x}", key_pair.address()));
+
+    let mut web3_instances = vec![];
+    let event_handlers = rpc_event_handlers_for_eth_transport(ctx, ticker.to_string());
+    for node in &nodes {
+        let mut transport = Web3Transport::with_event_handlers(vec![node.clone()], event_handlers.clone());
+        transport.gui_auth_validation_generator = Some(GuiAuthValidationGenerator {
+            coin_ticker: ticker.to_string(),
+            secret: key_pair.secret().clone(),
+            address: my_address.clone(),
+        });
+        drop_mutability!(transport);
+
+        let web3 = Web3::new(transport);
+        let version = match web3.web3().client_version().compat().await {
+            Ok(v) => v,
+            Err(e) => {
+                error!("Couldn't get client version for url {}: {}", node.uri, e);
+                continue;
+            },
+        };
+        web3_instances.push(Web3Instance {
+            web3,
+            is_parity: version.contains("Parity") || version.contains("parity"),
+        })
+    }
+
+    if web3_instances.is_empty() {
+        return Err(
+            EthActivationV2Error::UnreachableNodes("Failed to get client version for all nodes".to_string()).into(),
+        );
+    }
+
+    let mut transport = Web3Transport::with_event_handlers(nodes, event_handlers);
+    transport.gui_auth_validation_generator = Some(GuiAuthValidationGenerator {
+        coin_ticker: ticker.to_string(),
+        secret: key_pair.secret().clone(),
+        address: my_address,
+    });
+    drop_mutability!(transport);
+
+    let web3 = Web3::new(transport);
+
+    let (coin_type, decimals) = match protocol {
+        CoinProtocol::ETH => (EthCoinType::Eth, 18),
+        CoinProtocol::ERC20 {
+            platform,
+            contract_address,
+        } => {
+            let token_addr = valid_addr_from_str(&contract_address).map_err(EthActivationV2Error::InternalError)?;
+            let decimals = match conf["decimals"].as_u64() {
+                None | Some(0) => get_token_decimals(&web3, token_addr)
+                    .await
+                    .map_err(EthActivationV2Error::InternalError)?,
+                Some(d) => d as u8,
+            };
+            (EthCoinType::Erc20 { platform, token_addr }, decimals)
+        },
+        _ => {
+            return Err(EthActivationV2Error::InvalidPayload("Expect ETH or ERC20 protocol".to_string()).into());
+        },
+    };
+
+    // param from request should override the config
+    let required_confirmations = req
+        .required_confirmations
+        .unwrap_or_else(|| conf["required_confirmations"].as_u64().unwrap_or(1))
```

Comment:
Could you please add constant for this magic number and others?

Reply:
Done